grai-integrations/source-mysql/src/grai_source_mysql/loader.py

```diff
@@ -0,0 +1,206 @@
+        )
+        return [Table(**result, namespace=self.namespace) for result in res]
+
+    def get_columns(self, table: Table) -> List[Column]:
+        """
+        Creates and returns a list of dictionaries for the specified
+        schema.table in the database connected to.
+        """
+
+        query = f"""
+            SELECT column_name, data_type, is_nullable, column_default
+            FROM information_schema.columns
+            WHERE table_schema = '{table.table_schema}'
+            AND table_name = '{table.name}'
+            ORDER BY ordinal_position
+        """
+
+        res = (
+            {k.lower(): v for k, v in result.items()}
+            for result in self.query_runner(query)
+        )
+
+        addtl_args = {
+            "namespace": table.namespace,
+            "schema": table.table_schema,
+            "table": table.name,
+        }
+        return [Column(**result, **addtl_args) for result in res]
+
+    def get_foreign_keys(self) -> List[Edge]:
+        """This needs to be tested / evaluated
+        :param connection:
+        :param table:
+        :return:
+        """
+        # Only need constraint_types == 'f' for foreign keys but the others might be useful someday.
+        # Removed schemas, no schemas in mysql
+        query = """
+        select
+            tc.CONSTRAINT_NAME as constraint_name,
+            case when tc.CONSTRAINT_TYPE='PRIMARY KEY' then 'p'
+                when tc.CONSTRAINT_TYPE='FOREIGN KEY' then 'f' end as constraint_type,
+            tc.TABLE_schema as "self_schema",
+            tc.TABLE_NAME as "self_table",
+            GROUP_CONCAT(kcu.COLUMN_NAME SEPARATOR ',') "self_columns",
+            kcu.TABLE_schema as "foreign_schema",
+            kcu.REFERENCED_TABLE_NAME as "foreign_table",
+            GROUP_CONCAT(kcu.REFERENCED_COLUMN_NAME SEPARATOR ',') as "foreign_columns",
+            CONCAT(tc.CONSTRAINT_TYPE, ' (', GROUP_CONCAT(kcu.COLUMN_NAME), ')', (case when tc.CONSTRAINT_TYPE = 'FOREIGN KEY' then CONCAT(' REFERENCES ',kcu.REFERENCED_TABLE_NAME,'(',GROUP_CONCAT(kcu.REFERENCED_COLUMN_NAME),')') else '' end) ) as "definition"
+        from
+            information_schema.TABLE_CONSTRAINTS tc
+        join information_schema.KEY_COLUMN_USAGE kcu on
+            (kcu.TABLE_SCHEMA = tc.TABLE_SCHEMA
+            or kcu.REFERENCED_TABLE_SCHEMA = tc.TABLE_SCHEMA)
+            and (kcu.TABLE_NAME = tc.TABLE_NAME
+            or kcu.REFERENCED_TABLE_NAME = tc.TABLE_NAME)
+            and kcu.CONSTRAINT_NAME = tc.CONSTRAINT_NAME
+        group by
+            tc.CONSTRAINT_NAME,
+            tc.CONSTRAINT_TYPE,
+            tc.TABLE_NAME,
+            tc.TABLE_schema,
+            kcu.TABLE_schema,
+            kcu.REFERENCED_TABLE_NAME
+        """
+        addtl_args = {
+            "namespace": self.namespace,
+        }
+
+        res = self.query_runner(query)
+
+        for item in res:
+            item["self_columns"] = (
```

Comment:
@TonyLLondon I haven't had the opportunity to verify the changes on lines 169 through 176 are semantically correct but there's an edge case which occurs when the database is empty where the value in `self-columns` and `foreign_columns` is None rather than a string. The None value was in turn causing `split` to error.


I've modified it so that None passes through an empty list which _looked_ like the original intent.

Reply:
Yep that makes sense to me